### PR TITLE
React version

### PR DESCRIPTION
**Pull request by MeisamMahdian**
- React version
- Support of usage in list view
- New "button class" property

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+export default declare(`${widgetName}.widget.${widgetName}`, [_widgetBase], {
+  constructor() {
+    this.subscriptionHandles = [];
+    this.keyword = '';
+  },
+  postCreate() {
+    console.debug(`${this.id} >> postCreate`);
+  },
+  update(contextObject: mendix.lib.MxObject, callback: () => void) {
```
Comment (iannbing) [inline src/index.tsx:12]: `contextObject` is not used? Consider removing it?

File: src/index.webmodeler.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+
+import { LinkButton } from './components/LinkButton';
+import styleAsString from './style/style.scss';
+
+interface PreviewProps {
+  style: string;
+  class: string;
+  [key: string]: any;
```
Comment (iannbing) [inline src/index.webmodeler.tsx]: In the code below, it already shows that `props` has `labelCaption`, `linkTitle`, `renderType`, etc. Why not add them to the definition? Then we can get rid of the evil `any`

File: src/@typings/Widget.d.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export interface Widget {
+  mxform: mxui.lib.form._FormBase;
+  mxObject: mendix.lib.MxObject;
+  style?: React.CSSProperties;
```
Comment (iannbing) [inline src/@typings/Widget.d.ts]: Nice! I didn't know there's such type 👍

File: src/components/Link.tsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import classNames from 'classnames';
+import React from 'react';
+
+export interface LinkProps {
+  label: string;
+  onClickAction: (event: React.MouseEvent<HTMLElement>) => void;
+  iconUrl: string;
+  className?: string;
```
Comment (iannbing) [inline src/components/Link.tsx]: In `LinkButton`, `className` is not provided. So `className` will always be `undefined`.

File: src/components/LinkButton.tsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        iconUrl={props.iconUrl}
+        tooltip={props.tooltip}
+        className={props.buttonClass}
+        tabIndex={props.tabIndex}
+        bootstrapStyle={props.buttonStyle}
+      />
+    );
+  if (props.renderType === 'link')
+    return (
+      <Link
```
Comment (iannbing) [inline src/components/LinkButton.tsx:32]: `Link` can accept `className`, but it is not provided. Maybe pass `props.buttonClass` to `<Link />`?

File: src/components/LinkButton.tsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+import { Button } from './Button';
+import { Link } from './Link';
+
+export interface LinkButtonProps {
+  labelCaption: string;
+  tooltip: string;
+  iconUrl: string;
+  renderType: string;
+  style?: React.CSSProperties;
```
Comment (iannbing) [inline src/components/LinkButton.tsx]: `style` is not used anywhere.

File: src/components/LinkButton.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  className?: string;
```
Comment (iannbing) [inline src/components/LinkButton.tsx]: `className` is not used.

File: src/components/LinkButton.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  buttonStyle: BootstrapStyle;
```
Comment (iannbing) [inline src/components/LinkButton.tsx]: `buttonStyle` sounds like inline styling, but it's not. How about `buttonType`?

File: src/components/LinkButton.tsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  if (props.renderType === 'button')
+    return (
+      <Button
+        label={props.labelCaption}
+        onClickAction={props.onClick}
+        iconUrl={props.iconUrl}
+        tooltip={props.tooltip}
+        className={props.buttonClass}
+        tabIndex={props.tabIndex}
+        bootstrapStyle={props.buttonStyle}
```
Comment (iannbing) [inline src/components/LinkButton.tsx]: `bootstrapStyle` also sounds like inline styling. Would it be better to name it something like `bootstrapButtonType`?

File: src/components/__tests__/LinkButton.test.tsx [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    const props = {
+      onClick: handleClick,
+      className: 'awsomeClass',
+      tabIndex: 2,
+      style: sampleStyle,
+      labelCaption: 'labelCaption',
+      tooltip: 'buttonTitle',
+      renderType: 'button',
+      iconUrl: 'icon',
+      buttonstyle: 'info',
```
Comment (iannbing) [inline src/components/__tests__/LinkButton.test.tsx]: Should be camelcase here.

File: src/components/__tests__/LinkButton.test.tsx [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    const props = {
+      onClick: handleClick,
+      className: 'awsomeClass',
+      tabIndex: 2,
+      style: sampleStyle,
+      labelCaption: 'labelCaption',
+      tooltip: 'linkTitle',
+      renderType: 'link',
+      iconUrl: 'icon',
+      buttonstyle: '',
```
Comment (iannbing) [inline src/components/__tests__/LinkButton.test.tsx]: Should be camelcase here.

File: src/@typings/Widget.d.ts (hunk 3 above)
Comment (iannbing) [reply src/@typings/Widget.d.ts]: But, `style` is a `string` here :-)

File: src/container/index.tsx [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+  render(id, params, parent);
+};
+
+const render = (
+  id: string,
+  params: ClipboardTextCopyProps,
+  parent: HTMLElement
+) => {
+  const props: LinkButtonProps = {
+    style: parseStyle(params.style as string),
```
Comment (iannbing) [inline src/container/index.tsx]: `params.style` is indeed a `string`, but `Widget.d.ts` says it's a `React.CSSProperties`. If you fix `Widget.d.ts`, you don't need to assert its type here.

File: src/container/index.tsx [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    showError(targetClassName, id, parent);
+    return;
+  }
+  const text = !!(target as HTMLInputElement).value
+    ? (target as HTMLInputElement).value
+    : target.innerText;
+  if (!!text) copyToClipboard(text);
+};
+
+const findElement = (parent: HTMLElement, targetClassName: string) => {
```
Comment (iannbing) [inline src/container/index.tsx]: It is declared before being used. Could you move it to top?

File: src/container/index.tsx [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+  let target =
```
Comment (iannbing) [inline src/container/index.tsx]: `const` is better since it's not re-assigned.

File: src/container/index.tsx [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+) => {
+  const target = findElement(document.body, params.targetClassName);
+  if (!target) {
+    showError(params.targetClassName, id, parent);
+    return;
+  }
+  render(id, params, parent);
+};
+
+const render = (
```
Comment (iannbing) [inline src/container/index.tsx]: This is also declared before being used.

File: src/container/index.tsx [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+      event.preventDefault();
+      event.stopPropagation();
+      event.nativeEvent.stopImmediatePropagation();
+      return false;
+    },
+  };
+  ReactDOM.render(<LinkButton {...props} />, parent);
+};
+
+const handleClick = (
```
Comment (iannbing) [inline src/container/index.tsx:20]: This is also declared before being used.

File: src/container/index.tsx [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+    targetParent = targetParent.parentElement;
+    target = findElement(targetParent, targetClassName);
+  }
+
+  if (!target) {
+    showError(targetClassName, id, parent);
+    return;
+  }
+  const text = !!(target as HTMLInputElement).value
```
Comment (iannbing) [inline src/container/index.tsx]:
```
const text = (target as HTMLInputElement).value || target.innerText;
```

File: src/container/index.tsx [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+    ? (target as HTMLInputElement).value
+    : target.innerText;
+  if (!!text) copyToClipboard(text);
```
Comment (iannbing) [inline src/container/index.tsx:39]: Also show error when there's no text?

File: src/index.webmodeler.tsx [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+  style,
+  class: className,
+  ...props
+}: PreviewProps) => {
+  const nextProps = {
+    onClick: (event: React.MouseEvent<HTMLElement>) => {
+      event.preventDefault();
+      event.stopPropagation();
+      event.nativeEvent.stopImmediatePropagation();
+      return false;
```
Comment (iannbing) [inline src/index.webmodeler.tsx]: Why does it have to return `false`?

**Review by iannbing — CHANGES_REQUESTED** [review]
In `widget.config.xml`, it doesn't read data from `package.json`. For example, `widgetFriendlyName` and `description`. Please read `webpack.dev.js` and `webpack.prod.js`.

File: src/index.tsx (hunk 1 above)
Comment (MeisamMahdian) [reply src/index.tsx:12]: It's from Dojo. I can't remove it.

File: src/index.webmodeler.tsx (hunk 2 above)
Comment (MeisamMahdian) [reply src/index.webmodeler.tsx]: Done

File: src/@typings/Widget.d.ts (hunk 3 above)
Comment (MeisamMahdian) [reply src/@typings/Widget.d.ts]: Fixed

File: src/components/Link.tsx (hunk 4 above)
Comment (MeisamMahdian) [reply src/components/Link.tsx]: Removed

File: src/components/LinkButton.tsx (hunk 5 above)
Comment (MeisamMahdian) [reply src/components/LinkButton.tsx:32]: Removed

File: src/components/LinkButton.tsx (hunk 6 above)
Comment (MeisamMahdian) [reply src/components/LinkButton.tsx]: Removed

File: src/components/LinkButton.tsx (hunk 8 above)
Comment (MeisamMahdian) [reply src/components/LinkButton.tsx]: Done

File: src/components/LinkButton.tsx (hunk 9 above)
Comment (MeisamMahdian) [reply src/components/LinkButton.tsx]: Done

File: src/components/__tests__/LinkButton.test.tsx (hunk 11 above)
Comment (MeisamMahdian) [reply src/components/__tests__/LinkButton.test.tsx]: Fixed

File: src/components/LinkButton.tsx (hunk 7 above)
Comment (MeisamMahdian) [reply src/components/LinkButton.tsx]: Removed

File: src/components/__tests__/LinkButton.test.tsx (hunk 10 above)
Comment (MeisamMahdian) [reply src/components/__tests__/LinkButton.test.tsx]: Fixed

File: src/container/index.tsx (hunk 12 above)
Comment (MeisamMahdian) [reply src/container/index.tsx]: Removed

File: src/container/index.tsx (hunk 13 above)
Comment (MeisamMahdian) [reply src/container/index.tsx]: Done

File: src/container/index.tsx (hunk 14 above)
Comment (MeisamMahdian) [reply src/container/index.tsx]: Fixed

File: src/container/index.tsx (hunk 15 above)
Comment (MeisamMahdian) [reply src/container/index.tsx]: Done

File: src/container/index.tsx (hunk 16 above)
Comment (MeisamMahdian) [reply src/container/index.tsx:20]: Done

File: src/container/index.tsx (hunk 17 above)
Comment (MeisamMahdian) [reply src/container/index.tsx]: Done

File: src/container/index.tsx (hunk 18 above)
Comment (MeisamMahdian) [reply src/container/index.tsx:39]: I don't remember any environment that shows error for copying empty things. It shows error if it couldn't find the input text.

File: src/index.webmodeler.tsx (hunk 19 above)
Comment (MeisamMahdian) [reply src/index.webmodeler.tsx]: It was an old habit of stopPropagation in DOM events. But here in React wrapper its irrelevant.

**Comment by MeisamMahdian** [comment]
> In `widget.config.xml`, it doesn't read data from `package.json`. For example, `widgetFriendlyName` and `description`. Please read `webpack.dev.js` and `webpack.prod.js`.

Fixed that.

**Review by iannbing — APPROVED** [review]

**Pull request merged**